scalding-core/src/main/scala/com/twitter/scalding/JobClassFinder.scala

```diff
@@ -0,0 +1,55 @@
+package com.twitter.scalding
+
+import com.twitter.scalding.typed.CoGroupable
+import scala.reflect.runtime.universe
+
+object JobClassFinder {
+  private val baseContainers = List(
+    classOf[Execution[_]],
+    classOf[TypedPipe[_]],
+    classOf[TypedSink[_]],
+    classOf[CoGroupable[_, _]],
+    classOf[KeyedList[_, _]]
+  )
+  /**
+    * Reflect over a scalding job to try and identify types it uses so they can be tokenized by cascading.
+    * Since scala reflection is broken with the Hadoop InterfaceAudiance annotation (see
+    * https://issues.scala-lang.org/browse/SI-10129), we can't iterate over scalaType.members, so we instead use java
+    * reflection to iterate over fields to find the ones we care about, and then look those up in scala reflection to
+    * find the full un-erased type signatures, and try to find types from those.
+    *
+    * Note: this not guaranteed to find every used type. Eg, it can't find types used in a step that isn't
+    * referred to in a field
+    */
+  def findUsedClasses(jobClazz: Class[_ <: Job]): Set[Class[_]] = {
```

Comment:
Also, could we possibly also walk the methods and look at the input and return types? This could cover all the cases perhaps?